### DISCUSSION TITLE
Adding heic:lossless option for lossless HEIC/AVIF encoding

### DISCUSSION THREAD

**Discussion by dooman87** [ImageMagick/ImageMagick#3076] [Development] [answered]
Hi there,

I'd like to add a new option for lossless encoding when using HEIC/AVIF output format. I think we could make it similar to webp, like `-define heic:lossless=true`. Do you think it's a good idea and could it be useful? 

I'm happy to create a PR if you think it might work.

**Comment by urban-warrior** [comment, marked as answer]
Try adding -quality 100 to your command-line.  Does that produce what you need?  If not, feel free to post a PR.

**Reply by dooman87** [reply]
Thanks, quality=100 is not the same as lossless in libheif. I will create a PR shortly. Thanks for a quick reply!

**Comment by urban-warrior** [comment]
In your PR, trigger lossless encoding based on quality == 100.  That is how we do it in JPEG, for example.

**Reply by dooman87** [reply]
Thanks, Cristy. Just wondering what would be a pros/cons of using quality over a separate `define` flag? I can see that there is already `heic:preserve-orientation` define in heic coder. I was also thinking if it was possible to add a separate page for `heic` defines similar to [webp](https://imagemagick.org/script/webp.php)

**Comment by urban-warrior** [comment]
We already use a quality of 100 for other coders, certainly JPEG and couple of others.  So the expectation for our ImageMagick users is that a quality of 100 means lossless.  A principle of development is orthogonality-- that is one way to do one thing.  Using a define for HEIC and quality of 100 for other coders would violate this important design principle.

Having a separate web page for HEIC would be grand.  As you know, we have a repo for the web pages.  Feel free to add a pull request in the repo with an HEIC page describing what it is, how to effectively use it, and any defines that might influence its usage.  The ImageMagick user community will benefit from your efforts so a quick thank you for any contributions you submit.

**Reply by dooman87** [reply]
Thanks Cristy, the above makes perfect sense.

**Comment by Jmuccigr** [comment]
Was this ever implemented?

**Reply by dlemstra** [reply]
This was added in this pull request: #3116.

**Reply by Jmuccigr** [reply]
Thanks. I just converted a tiff to heic using `-quality 100` and got an image with differences to the original tiff, according to `compare -metric mae -verbose`, which gives:

```
  Channel distortion: MAE
    red: 187.587 (0.00286239)
    green: 102.677 (0.00156674)
    blue: 280.387 (0.00427843)
    all: 190.217 (0.00290252)
```

The profiles are different, so maybe that's it? See #1947 .

**Comment by snibgo** [comment]
What are the bit-depths of your TIFF and HEIC? For example, if they are 16 and 8, you can't expect an accuracy greater than about 1/256, or 0.004.

**Reply by Jmuccigr** [reply]
`identify -format "%[bit-depth]"` reports 8 for both.

**Comment by snibgo** [comment]
What colour models are used? For example, if your TIFF model is RGB but the HEIC is YCbCr (exiftool will tell you this), then the HEIC will be converted back to RGB for the comparison. Again, we can't expect an accuracy better than about 1 part in 256 for 8-bit values.

**Reply by Jmuccigr** [reply]
If I'm looking in the right place, exiftool reports the same thing for both:

`Profile Description             : Adobe RGB (1998)`

Also for the heic I get:

```
Colorspace: sRGB 
Type: TrueColor
```

tif has:

`Color Space Data                : RGB`

As a comparison, if I convert the tiff to png, `compare` reports 0 for each channel, and the output png looks like a very washed out version of the original, while the same file in the heic case is very red. (Not sure what that image is supposed to show, TBH.)

**Comment by snibgo** [comment]
Exiftool "Color Space Data" refers to the components for the embedded ICC colour profile. So the components of the "Adobe RGB (1998)" profile are RGB, as we might expect.

But a file with that embedded profile may have:
```
Components Configuration        : Y, Cb, Cr, -
```
When IM reads this file, the HEIC library will convert the pixel data from YCbCr to RGB. And the opposite for writing.

> if I convert the tiff to png, compare reports 0 for each channel, and the output png looks like a very washed out version of the original,

A simple conversion that reads the tiff and writes a png should include the embedded "Adobe RGB (1998)" profile. If your file viewer correctly process the profiles, the images should look the same. But the png looks more washed out than the tiff, so either the png doesn't contain the profile, or the viewer is ignoring it.

**Reply by Jmuccigr** [reply]
Sorry, I meant that the output png from the compare command looks like a washed out version of the original (or a very red version if I compare the heic to the tiff). The converted png looks identical to the original tiff on visual inspection, and if I use Graphic Converter to do a difference between the two files, it finds none, which is not the case for the heic.

exiftool does indeed report this info for the heic:

`Components Configuration        : Y, Cb, Cr, -`

So I guess the question is whether it's possible to avoid this round-trip through YCbCr (if I'm understanding all this correctly).
